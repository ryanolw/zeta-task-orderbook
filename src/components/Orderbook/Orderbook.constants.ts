export const COLUMN_WIDTHS = ['50%', '50%'];
