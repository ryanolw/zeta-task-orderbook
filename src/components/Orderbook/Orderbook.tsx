import { useInterval } from '@chakra-ui/react';
import { useEffect, useState } from 'react';
import type { OrderBookData, OrderbookProps } from './Orderbook.types';
import dataAdapters from './adapters';
import OrderbookBase from './components/OrderbookBase';

export default function Orderbook({
  apiUrl,
  dataAdapter = 'zeta',
  pollingIntervalMs = 5000,
}: OrderbookProps) {
  const [data, setData] = useState<OrderBookData | null>(null);
  const [error, setError] = useState<Error | null>(null);
  const fetchOrderData = async () => {
    const orderDataFetcher =
      typeof dataAdapter === 'function'
        ? dataAdapter
        : dataAdapters[dataAdapter];
    try {
      const orderData: OrderBookData = await orderDataFetcher(apiUrl);
      setData(orderData);
      setError(null);
    } catch (err) {
      setError(err as Error);
    }
  };

  useEffect(() => {
    // useInterval hook starts after initial delay, so manually trigger a fetch right after mounting
    fetchOrderData();
  }, []);

  useInterval(() => {
    fetchOrderData();
  }, pollingIntervalMs);

  return <OrderbookBase data={data} error={error} />;
}
