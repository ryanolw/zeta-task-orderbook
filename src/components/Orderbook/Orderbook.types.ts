export type OrderBookData = { bids: Order[]; asks: Order[] };

type DataAdapter = (apiUrl: string) => Promise<OrderBookData>;

export type DataAdapters = Record<'zeta', DataAdapter>;

export type Order = {
  price: number;
  size: number;
};
export type EnrichedOrderProp = {
  cumulativeTotalPct: number;
  orderTotal: number;
  cumulativeTotal: number;
  price: number;
  size: number;
};

export type OrderbookProps = {
  /**
   * API endpoint to retrieve the orderbook data.
   */
  apiUrl: string;

  /**
   * Name of the whitelisted dataAdapter to use. The data adapter is responsible for transforming data retrieved from `apiUrl` to a structure that is understood by the Orderbook component. default: "zeta"
   * A custom data adapter can also be used by passing in a function. The function will receive an `apiUrl` argument and should return an object that satisfies the `OrderBookData` type.
   */
  dataAdapter: keyof DataAdapters | DataAdapter;

  /**
   * Interval (in milliseconds) to poll `apiUrl` for new data. Default: 5000
   */
  pollingIntervalMs?: number;
};
