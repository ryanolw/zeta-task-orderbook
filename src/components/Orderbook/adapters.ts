import { DataAdapters, OrderBookData } from './Orderbook.types';

const dataAdapters: DataAdapters = {
  zeta: async (apiUrl: string) => {
    const result: { orderbooks: OrderBookData[] } = await fetch(apiUrl).then(
      (res) => res.json()
    );
    return result.orderbooks[0];
  },
};

export default dataAdapters;
