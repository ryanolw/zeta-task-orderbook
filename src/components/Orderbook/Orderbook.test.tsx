import { render, waitForElementToBeRemoved } from '@testing-library/react';
import Orderbook from './Orderbook';
import dataAdapters from './adapters';

jest.mock('./adapters');

describe('Orderbook', () => {
  test('should match snapshot when rendering order data using whitelisted data adapter', async () => {
    (dataAdapters.zeta as jest.Mock).mockReturnValue(sampleData);
    const res = render(<Orderbook dataAdapter={({ zeta }) => zeta('foo')} />);
    await waitForElementToBeRemoved(await res.findByText('Loading...'));
    expect(res.asFragment()).toMatchSnapshot();
  });

  test('should match snapshot when rendering order data using custom data adapter', async () => {
    const res = render(<Orderbook dataAdapter={() => sampleData} />);
    await waitForElementToBeRemoved(await res.findByText('Loading...'));
    expect(res.asFragment()).toMatchSnapshot();
  });
});

const sampleData = {
  bids: [
    { price: 18.7201, size: 168.895 },
    { price: 18.717, size: 10.69 },
    { price: 18.7163, size: 844.473 },
    { price: 18.7089, size: 2251.928 },
    { price: 18.708, size: 53.45 },
    { price: 18.7014, size: 2529.446 },
    { price: 18.698, size: 267.26 },
    { price: 18.689, size: 267.26 },
    { price: 18.642, size: 801.79 },
    { price: 18, size: 0.084 },
    { price: 15, size: 0.05 },
    { price: 2.01, size: 3 },
    { price: 2, size: 1 },
    { price: 1.3, size: 1 },
    { price: 0.01, size: 1.66 },
  ],
  asks: [
    { price: 18.7303, size: 151.701 },
    { price: 18.734, size: 758.507 },
    { price: 18.7415, size: 2022.685 },
    { price: 18.749, size: 1142.531 },
    { price: 19, size: 5 },
    { price: 19.25, size: 5 },
    { price: 19.5, size: 5 },
    { price: 20, size: 5 },
    { price: 21, size: 5 },
    { price: 24.5, size: 0.5 },
    { price: 25, size: 0.5 },
    { price: 26, size: 0.38 },
  ],
};
