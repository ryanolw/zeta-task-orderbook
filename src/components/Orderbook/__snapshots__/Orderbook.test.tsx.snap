// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Orderbook should match snapshot when rendering order data 1`] = `
<DocumentFragment>
  <div
    class="css-1usl3sb"
  >
    <div
      class="chakra-stack css-1ne2chk"
    >
      <p
        class="chakra-text css-11ze7cv"
      >
        Price
      </p>
      <p
        class="chakra-text css-11ze7cv"
      >
        Size
      </p>
    </div>
    <div
      class="chakra-stack css-1gx94zz"
    >
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-1ogj3e"
        />
        <p
          class="chakra-text css-9fnj8l"
        >
          18.730
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          151.701
        </p>
      </div>
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-1tqpelg"
        />
        <p
          class="chakra-text css-9fnj8l"
        >
          18.734
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          758.507
        </p>
      </div>
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-1bg6yau"
        />
        <p
          class="chakra-text css-9fnj8l"
        >
          18.741
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          2022.685
        </p>
      </div>
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-gmalf9"
        />
        <p
          class="chakra-text css-9fnj8l"
        >
          18.749
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          1142.531
        </p>
      </div>
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-ftigz9"
        />
        <p
          class="chakra-text css-9fnj8l"
        >
          19.000
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          5.000
        </p>
      </div>
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-vpccyc"
        />
        <p
          class="chakra-text css-9fnj8l"
        >
          19.250
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          5.000
        </p>
      </div>
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-5a5tht"
        />
        <p
          class="chakra-text css-9fnj8l"
        >
          19.500
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          5.000
        </p>
      </div>
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-1i4wv4x"
        />
        <p
          class="chakra-text css-9fnj8l"
        >
          20.000
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          5.000
        </p>
      </div>
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-fxafsu"
        />
        <p
          class="chakra-text css-9fnj8l"
        >
          21.000
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          5.000
        </p>
      </div>
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-n5msg4"
        />
        <p
          class="chakra-text css-9fnj8l"
        >
          24.500
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          0.500
        </p>
      </div>
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-1stdgjf"
        />
        <p
          class="chakra-text css-9fnj8l"
        >
          25.000
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          0.500
        </p>
      </div>
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-1k77zqf"
        />
        <p
          class="chakra-text css-9fnj8l"
        >
          26.000
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          0.380
        </p>
      </div>
    </div>
    <div
      class="chakra-stack css-165casq"
    >
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-1oho7yd"
        />
        <p
          class="chakra-text css-rb85yi"
        >
          18.720
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          168.895
        </p>
      </div>
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-gyvep2"
        />
        <p
          class="chakra-text css-rb85yi"
        >
          18.717
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          10.690
        </p>
      </div>
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-1sx24n2"
        />
        <p
          class="chakra-text css-rb85yi"
        >
          18.716
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          844.473
        </p>
      </div>
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-6e7ghd"
        />
        <p
          class="chakra-text css-rb85yi"
        >
          18.709
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          2251.928
        </p>
      </div>
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-nw9n8c"
        />
        <p
          class="chakra-text css-rb85yi"
        >
          18.708
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          53.450
        </p>
      </div>
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-pehgtx"
        />
        <p
          class="chakra-text css-rb85yi"
        >
          18.701
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          2529.446
        </p>
      </div>
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-13cpeq9"
        />
        <p
          class="chakra-text css-rb85yi"
        >
          18.698
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          267.260
        </p>
      </div>
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-1wile65"
        />
        <p
          class="chakra-text css-rb85yi"
        >
          18.689
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          267.260
        </p>
      </div>
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-1ewvd5i"
        />
        <p
          class="chakra-text css-rb85yi"
        >
          18.642
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          801.790
        </p>
      </div>
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-kmz5gy"
        />
        <p
          class="chakra-text css-rb85yi"
        >
          18.000
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          0.084
        </p>
      </div>
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-jcpbrj"
        />
        <p
          class="chakra-text css-rb85yi"
        >
          15.000
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          0.050
        </p>
      </div>
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-4ea3pr"
        />
        <p
          class="chakra-text css-rb85yi"
        >
          2.010
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          3.000
        </p>
      </div>
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-1eme6k2"
        />
        <p
          class="chakra-text css-rb85yi"
        >
          2.000
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          1.000
        </p>
      </div>
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-w93jsg"
        />
        <p
          class="chakra-text css-rb85yi"
        >
          1.300
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          1.000
        </p>
      </div>
      <div
        class="chakra-stack css-p72o2v"
      >
        <div
          class="css-1xli5rx"
        />
        <p
          class="chakra-text css-rb85yi"
        >
          0.010
        </p>
        <p
          class="chakra-text css-11ze7cv"
        >
          1.660
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
